### DISCUSSION TITLE
Is it a goal for Gleam to work entirely without Erlang some day?

### DISCUSSION THREAD

**Discussion by johannesvollmer** [gleam-lang/gleam#3138] [Ideas & suggestions]
The Gleam release seems to have sparked lot of excitement in the frontend world. I'm just trying it out, and was delighted to find the Gleam core to work without Erlang. I currently don't have Erlang installed, only NodeJS. I can imagine many frontend folks are in the same situation as me.  I have never worked with Erlang and don't plan to, and I have a little uncertainty concerning whether Gleam is the right language for me.

I observed the following points of uncertainty and friction:
- hex packages seem to be mixed across targets - how can I know if a package supports my javascript target?
- default target is erlang, and needs to be changed manually after running `gleam new`, as it doesn't offer a default way to create javascript projects
- as an experienced elm person, I was thrilled to try out lustre. here's how it went:
   1. as gleam apparently doesn't compile to a single javascript file, it seems a more complex build process is required than simply running `gleam build` and referencing the javascript file in the html (or at least I couldn't find the documentation, if it exists).
   2. lustre therefore has a command which does all of that and more, `gleam run -m lustre/dev start`
   3. as it turns out, that script requires erlang, so I installed that, which was rather easy
   4. as it turns out, that script also requires rebar3, and installing that seems like a complete mystery to me

The fact that the core Gleam runs without Erlang was pretty amazing to see. But it many other places, Erlang seems like a strict requirement. Has someone of you made a similar experience?

Are front-end people without Erlang experience the target audience of Gleam? Is it an explicit Goal of Gleam to only work with a full-blown Erlang installation? Or will javascript some day be a standalone target? Or are there maybe even plans about that?

**Comment by lpil** [comment]
Hello!

> hex packages seem to be mixed across targets - how can I know if a package supports my javascript target?

This is something we will have in future. The compiler does gather and publish this information, and we're waiting for a good deal of the packages to be published with it.

>    default target is erlang, and needs to be changed manually after running gleam new, as it doesn't offer a default way to create javascript projects

We may have something akin to templates in future but there are no proposals for how they might work today. If you have ideas we are very interested!

>    as an experienced elm person, I was thrilled to try out lustre. here's how it went:
>  as gleam apparently doesn't compile to a single javascript file, it seems a more complex build process is required than simply running gleam build and referencing the javascript file in the html (or at least I couldn't find the documentation, if it exists).
> lustre therefore has a command which does all of that and more, gleam run -m lustre/dev start

These days you do not need to bundle JavaScript to run or deploy it, so you could opt to not use Lustre's tooling if you prefer.

>  as it turns out, that script requires erlang, so I installed that, which was rather easy
>  as it turns out, that script also requires rebar3, and installing that seems like a complete mystery to me

I can't speak for the Lustre maintainers but I suspect they picked Erlang over JavaScript as Gleam programmers are more likely to have that installed. Documentation can be found here: <https://gleam.run/getting-started/installing/>

         

> The fact that the core Gleam runs without Erlang was pretty amazing to see. But it many other places, Erlang seems like a strict requirement. Has someone of you made a similar experience?

There is no strict Erlang requirement to use Gleam. You may use any Erlang or JS runtime, and we have built in support for BEAM, Deno, Bun, and NodeJS.

> Are front-end people without Erlang experience the target audience of Gleam?

Yes. These people are very common in the Gleam community.

> Is it an explicit Goal of Gleam to only work with a full-blown Erlang installation?

No.

> Or will javascript some day be a standalone target? Or are there maybe even plans about that?

It always has been.

**Reply by johannesvollmer** [reply]
awesome, thanks for your detailed response. it's reassuring:) I'm excited for my further gleam journey
